File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    answer = 0
    dp = [[0 for _ in range(m)] for _ in range(n)]
    dp[0][0] = 1
    
    puddle_set = set()
    
    for cur_puddle_m, cur_puddle_n in puddles:
        puddle_set.add((cur_puddle_m, cur_puddle_n))
    
    for cur_m in range(m):
        for cur_n in range(n):
            for next_m, next_n in [[cur_m + 1, cur_n], [cur_m, cur_n + 1]]:
                if 0 <= next_m < m and 0 <= next_n < n and (next_m + 1, next_n + 1) not in puddle_set:
                    dp[next_n][next_m] = (dp[next_n][next_m] + dp[cur_n][cur_m]) % 1000000007
                    
    return dp[n - 1][m - 1]

File: 프로그래머스/3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(int m, int n, vector<vector<int>> puddles) {
    
    vector<vector<bool>> board(n, vector<bool>(m, false));
    for(int i = 0; i < puddles.size(); i++){
        int cur_m = puddles[i][0];
        int cur_n = puddles[i][1];
        board[cur_n - 1][cur_m - 1] = true;
    }
    vector<vector<int>> dp(n, vector<int>(m, 0));
    dp[0][0] = 1;
    
    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            if(board[i][j] || (i == 0 && j == 0))
                continue;
            
            int cur_path = 0;
            if(i > 0){
                cur_path += dp[i - 1][j];
            }
            
            if(j > 0){
                cur_path += dp[i][j - 1];
            }
            
            dp[i][j] = cur_path % 1000000007;
        }
    }
    
    return dp[n - 1][m - 1] % 1000000007;
}